### ISSUE TITLE
When is is safe to setFlushedLSN when write-in-chunks is enabled?

### ISSUE THREAD

**Issue by igouss** [eulerto/wal2json#168]
Is it safe to setFlushedLSN in the middle of the transaction? Would is restart from the first statement of the transaction?

If setFlushedLSN is called in the middle of the transaction and then connection is closed, would wal2json start streaming from the middle of the transaction on reconnect?

Should nextlsn be used when calling setFlushedLSN?

Also, is txId has 1-to-1 mapping to LSN? If not does wal2json start streaming from the beginning of LSN block? What is LSN block starts from the middle of the transaction?

(java, jdbc)

**Comment by eulerto** [comment]
My answers are based on the streaming replication protocol using tools such as `psql` and `pg_recvlogical`.

> Is it safe to setFlushedLSN in the middle of the transaction? Would is restart from the first statement of the transaction?
> If setFlushedLSN is called in the middle of the transaction and then connection is closed, would wal2json start streaming from the middle of the transaction on reconnect?
>
No. It should start from the beginning of the transaction if the connection is closed while processing the transaction. I don't know if pgjdbc behaves sanely. Better ask those questions to pgjdbc developers.

Bear in mind that write-in-chunks is an artifact to transmit long transactions. You should consider the format 2.

> Should nextlsn be used when calling setFlushedLSN?
>
No. nextlsn means next transaction. If you do that you are skipping the current transaction.

> Also, is txId has 1-to-1 mapping to LSN? If not does wal2json start streaming from the beginning of LSN block? What is LSN block starts from the middle of the transaction?
>
I don't understand what you mean by "1-to-1 mapping". If you think as a streaming, LSN represents the position where the operations starts (such as transactions, inserts, inserts into indexes, etc) -- length of each operation is also available.
If a connection is broken and you start again, wal2json will start from the beginning of the transaction which is represented by a LSN (`pg_replication_slots.confirmed_flush_lsn`).

**Comment by igouss** [comment]
So using v2 format and the example from the front page:

```json
{"action":"B","xid":554737,"timestamp":"2020-06-30 14:00:22.296699-04","lsn":"1/A451B118"}
{"action":"I","xid":554737,"timestamp":"2020-06-30 14:00:22.296699-04","lsn":"1/A451AD40","schema":"public","table":"table2_with_pk","columns":[{"name":"a","type":"integer","value":1},{"name":"b","type":"character varying(30)","value":"Backup and Restore"},{"name":"c","type":"timestamp without time zone","value":"2020-06-30 14:00:22.282816"}]}
{"action":"I","xid":554737,"timestamp":"2020-06-30 14:00:22.296699-04","lsn":"1/A451AE40","schema":"public","table":"table2_with_pk","columns":[{"name":"a","type":"integer","value":2},{"name":"b","type":"character varying(30)","value":"Tuning"},{"name":"c","type":"timestamp without time zone","value":"2020-06-30 14:00:22.282816"}]}
{"action":"I","xid":554737,"timestamp":"2020-06-30 14:00:22.296699-04","lsn":"1/A451AED8","schema":"public","table":"table2_with_pk","columns":[{"name":"a","type":"integer","value":3},{"name":"b","type":"character varying(30)","value":"Replication"},{"name":"c","type":"timestamp without time zone","value":"2020-06-30 14:00:22.282816"}]}
{"action":"D","xid":554737,"timestamp":"2020-06-30 14:00:22.296699-04","lsn":"1/A451AF70","schema":"public","table":"table2_with_pk","identity":[{"name":"a","type":"integer","value":1},{"name":"c","type":"timestamp without time zone","value":"2020-06-30 14:00:22.282816"}]}
{"action":"D","xid":554737,"timestamp":"2020-06-30 14:00:22.296699-04","lsn":"1/A451AFC0","schema":"public","table":"table2_with_pk","identity":[{"name":"a","type":"integer","value":2},{"name":"c","type":"timestamp without time zone","value":"2020-06-30 14:00:22.282816"}]}
{"action":"I","xid":554737,"timestamp":"2020-06-30 14:00:22.296699-04","lsn":"1/A451B078","schema":"public","table":"table2_without_pk","columns":[{"name":"a","type":"integer","value":1},{"name":"b","type":"numeric(5,2)","value":2.34},{"name":"c","type":"text","value":"Tapir"}]}
WARNING:  no tuple identifier for UPDATE in table "public"."table2_without_pk"
{"action":"C","xid":554737,"timestamp":"2020-06-30 14:00:22.296699-04","lsn":"1/A451B118"}
```

To ensure that the transaction is either re-sent if connection is dropped in the middle of transaction processing or if it is successfully processed, we should only setFlushedLSN to LSN from "action":"C" ?

**Comment by eulerto** [comment]
> To ensure that the transaction is either re-sent if connection is dropped in the middle of transaction processing or if it is successfully processed, we should only setFlushedLSN to LSN from "action":"C" ? Or we should use LSN from the previous transaction?

No. [checking the code...] format 1 has `nextlsn` that is the end of the current commit record + 1 which is the next transaction as a starting point. format 2 does not provide `nextlsn`. Ooops. That's a design flaw and I will implement it later.

**Comment by eulerto** [comment]
It was fixed in commit 8600d480de304ffe4c6a313b6fba1e411ab2e706.
